janus_core/processing/observables.py

```diff
@@ -138,9 +138,8 @@ def _set_components(self, components: list[str]):
         ValueError
             If any component is invalid.
         """
-        for component in components:
+        for component in self.allowed_components.keys() - components.keys():
             if component not in self.allowed_components:
-                component_names = list(self._components.keys())
                 raise ValueError(
                     f"'{component}' invalid, must be '{', '.join(component_names)}'"
                 )
```

Comment:
Blindly applying this isn't right.

You no longer need to check the `if`, as that's done by the operation. `component_names` is not just undefined and need to refer to the components which are not defined. 

```python
if any(self.allowed_components.keys() - components.keys()):
    raise ValueError(
                    f"'{','.join(self.allowed_components.keys() - components.keys())}' invalid, must be in '{', '.join(self.allowed_components)}'"
    )

Reply:
yes, unfortunately I did check the changes locally but then added them all together anyway... the code was odd due to the line length but this can be solved

Reply:
It's my problem really, I click "suggestion" because I'm too lazy to match the formatting and then it renders with "add this" even when it's invalid code. I should change the type to `Python`